### ISSUE TITLE
Error: P1002 database server was reached but timed out

### ISSUE THREAD

**Issue by s-kris** [prisma/prisma#6768]
## Bug description
I was running `prisma migrate dev` and ran into this error.

```
Error: P1002

The database server at `xx.xx.xx.xx`:`5432` was reached but timed out.

Please try again.

Please make sure your database server is running at `xx.xx.xx.xx`:`5432`.
```
I'm not using pgBouncer. This is my connection string: 
`postgres://<XXXXXXXXXX>/defaultdb?connection_limit=3&connection_timeout=30&pool_timeout=30`

I looked at #6053

## How to reproduce
This is a random error.

## Expected behavior
Primsa migrate happens as expected

## Environment & setup

- OS:  Mac OS 10.15.7 
- Database:  Postgres
- Node.js version:  15.14.0
- Prisma version:

```
prisma               : 2.21.2
@prisma/client       : 2.21.2
Current platform     : darwin
Query Engine         : query-engine e421996c87d5f3c8f7eeadd502d4ad402c89464d (at node_modules/@prisma/engines/query-engine-darwin)
Migration Engine     : migration-engine-cli e421996c87d5f3c8f7eeadd502d4ad402c89464d (at node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine : introspection-core e421996c87d5f3c8f7eeadd502d4ad402c89464d (at node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary        : prisma-fmt e421996c87d5f3c8f7eeadd502d4ad402c89464d (at node_modules/@prisma/engines/prisma-fmt-darwin)
Default Engines Hash : e421996c87d5f3c8f7eeadd502d4ad402c89464d
Studio               : 0.371.0
```

**Comment by janpio** [comment]
Can you do other things (Introspection, Queries) with this connection string?

**Comment by s-kris** [comment]
Interesting. 
For introspection, I'm getting `P1001 error`

Queries and Mutations work fine though.

**Comment by s-kris** [comment]
I deleted node_modules folder and reinstalled the packages. 
Then I ran prisma migrate, got this error

```
Error: P1017

Server has closed the connection.
```

I just ran the command again, it's worked.

**Comment by janpio** [comment]
This all points to your database server being only reachable and stable sporadically. Could that be possible?

**Comment by s-kris** [comment]
I don't think so, cause other queries and mutations work at the same time.

**Comment by janpio** [comment]
Can you summarize what works, and what does not?
All of these are using the same connection string setup?

**Comment by s-kris** [comment]
1. I'm using the same connection string for development and migrations.
2. Queries, mutations are working all the time. 
3. Migrations are throwing the above error randomly.

**Comment by pantharshit00** [comment]
@s-kris Can you please tell us where your database is hosted? As you can guess it would be quite hard to reproduce this without knowing all the details. 

Also, run `export DEBUG=*` before running migrate and paste the output here.

**Comment by s-kris** [comment]
Hi @pantharshit00
My dev-database is hosted on hetzner cloud.
I'll post the debug log when the error is thrown (it seems be random as of now)

**Comment by ChukaEbi** [comment]
Hi @pantharshit00 I am running into the same issue and it is happening consistently. Issue only started as of the last 5 minutes. I rolled back all changes in that timespan but the issue still persists

My DB is hosted on Supabase

Debug log below:

```
prisma:loadEnv project root found at /Users/ceb/projects/kana-api/package.json +0ms
  prisma:tryLoadEnv Environment variables loaded from /Users/ceb/projects/kana-api/.env +0ms
Environment variables loaded from .env
  prisma:engines using NAPI: false +0ms
  prisma:engines binaries to download query-engine, migration-engine, introspection-engine, prisma-fmt +0ms
Prisma schema loaded from prisma/schema.prisma
Datasource "db": PostgreSQL database "postgres", schema "public" at "db.xxxxxxx.supabase.co:5432"

  prisma:migrateEngine:rpc starting migration engine with binary: /Users/ceb/projects/kana-api/node_modules/@prisma/engines/migration-engine-darwin +0ms
  prisma:migrateEngine:rpc SENDING RPC CALL {"id":1,"jsonrpc":"2.0","method":"devDiagnostic","params":{"migrationsDirectoryPath":"/Users/ceb/projects/kana-api/prisma/migrations"}} +5ms
  prisma:migrateEngine:stderr May 03 17:08:57.119  INFO migration_engine: Starting migration engine RPC server git_hash="e421996c87d5f3c8f7eeadd502d4ad402c89464d" +0ms
  prisma:migrateEngine:stderr May 03 17:08:57.503  INFO quaint::single: Starting a postgresql connection. +384ms
  prisma:migrateEngine:stderr May 03 17:08:58.375  INFO DevDiagnostic:calculate_drift:sql_schema_from_migration_history: quaint::single: Starting a postgresql connection. +871ms
  prisma:migrateEngine:stderr May 03 17:09:03.093  INFO DevDiagnostic:validate_migrations:sql_schema_from_migration_history: quaint::single: Starting a postgresql connection. +5s
  prisma:migrate:dev {
  prisma:migrate:dev   devDiagnostic: '{\n  "action": {\n    "tag": "createMigration"\n  }\n}'
  prisma:migrate:dev } +0ms
  prisma:migrateEngine:rpc SENDING RPC CALL {"id":2,"jsonrpc":"2.0","method":"applyMigrations","params":{"migrationsDirectoryPath":"/Users/ceb/projects/kana-api/prisma/migrations"}} +9s
  prisma:migrateEngine:rpc {
  prisma:migrateEngine:rpc   jsonrpc: '2.0',
  prisma:migrateEngine:rpc   error: {
  prisma:migrateEngine:rpc     code: 4466,
  prisma:migrateEngine:rpc     message: 'An error happened. Check the data field for details.',
  prisma:migrateEngine:rpc     data: {
  prisma:migrateEngine:rpc       is_panic: false,
  prisma:migrateEngine:rpc       message: 'The database server at `db.xxxxxxx.supabase.co`:`5432` was reached but timed out.\n' +
  prisma:migrateEngine:rpc         '\n' +
  prisma:migrateEngine:rpc         'Please try again.\n' +
  prisma:migrateEngine:rpc         '\n' +
  prisma:migrateEngine:rpc         'Please make sure your database server is running at `db.mqwxclxnqubbhcssyjnf.supabase.co`:`5432`.\n',
  prisma:migrateEngine:rpc       meta: [Object],
  prisma:migrateEngine:rpc       error_code: 'P1002'
  prisma:migrateEngine:rpc     }
  prisma:migrateEngine:rpc   },
  prisma:migrateEngine:rpc   id: 2
  prisma:migrateEngine:rpc } +10s
Error: Error: P1002

The database server at `db.xxxxxxxx.supabase.co`:`5432` was reached but timed out.

Please try again.

Please make sure your database server is running at `db.xxxxxxxx.supabase.co`:`5432`.


    at Object.<anonymous> (/Users/ceb/projects/kana-api/node_modules/prisma/build/index.js:55736:26)
    at MigrateEngine.handleResponse (/Users/ceb/projects/kana-api/node_modules/prisma/build/index.js:55611:38)
    at LineStream.<anonymous> (/Users/ceb/projects/kana-api/node_modules/prisma/build/index.js:55696:18)
    at LineStream.emit (node:events:369:20)
    at LineStream.emit (node:domain:470:12)
    at addChunk (node:internal/streams/readable:313:12)
    at readableAddChunk (node:internal/streams/readable:288:9)
    at LineStream.Readable.push (node:internal/streams/readable:227:10)
    at LineStream._pushBuffer (/Users/ceb/projects/kana-api/node_modules/prisma/build/index.js:55451:19)
    at LineStream._transform (/Users/ceb/projects/kana-api/node_modules/prisma/build/index.js:55445:10)
```

### Environment & setup

OS: Mac OS 10.15.4 
Database: Postgres
Node.js version: v15.12.0
Prisma version:

```
prisma               : 2.21.2
@prisma/client       : 2.20.1
Current platform     : darwin
Query Engine         : query-engine e421996c87d5f3c8f7eeadd502d4ad402c89464d (at node_modules/@prisma/engines/query-engine-darwin)
Migration Engine     : migration-engine-cli e421996c87d5f3c8f7eeadd502d4ad402c89464d (at node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine : introspection-core e421996c87d5f3c8f7eeadd502d4ad402c89464d (at node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary        : prisma-fmt e421996c87d5f3c8f7eeadd502d4ad402c89464d (at node_modules/@prisma/engines/prisma-fmt-darwin)
Default Engines Hash : e421996c87d5f3c8f7eeadd502d4ad402c89464d
Studio               : 0.371.0
```

**Comment by janpio** [comment]
What command exactly are you running @ChukaEbi? What kind of migration (and how much) would you expect that command to create and execute?

**Comment by ChukaEbi** [comment]
@janpio I am running `npx prisma migrate dev` and I expected it to just add a uniqueness constraint to an existing field in a table.

**Comment by pantharshit00** [comment]
@ChukaEbi Are you able to reproduce this in the latest version?

**Comment by meotimdihia** [comment]
I am still geting this bug with 2.26.

**Comment by pantharshit00** [comment]
@meotimdihia Can you tell us where you are hosting your database?

**Comment by meotimdihia** [comment]
@pantharshit00 I don't get this error anymore -  Prisma 2.27.

**Comment by pantharshit00** [comment]
Closing due to inactivity

**Comment by larsniet** [comment]
Issue persists... Running Windows 10 and connected Prisma to PostgreSQL running on Ubuntu 20. 
```
Error: P1002

The database server at `37.97.209.87`:`5432` was reached but timed out.

Please try again.

Please make sure your database server is running at `37.97.209.87`:`5432`.

Context: Timed out trying to acquire a postgres advisory lock (SELECT pg_advisory_lock(72707369)). Elapsed: 10000ms. See https://pris.ly/d/migrate-advisory-locking for details.
```
Tried reinstalling node modules and undoing commits, nothing seems to work.

**Comment by gregorcee123** [comment]
@s-kris Did you resolve this, i'm still consistently getting this, database is supabase as well

**Comment by mathiasklenk** [comment]
Same, this is happening at random intervals and just annoying. I am also using Supabase...

**Comment by s-kris** [comment]
@gregorcee123 @mathiasklenk 

I'm on version 2.27. This error doesn't seem to happen. Atleast not in the last six months. 

My databases are on AWS RDS and Supabase(recent project). 

I should also mention that my prisma + nextjs setup is a bit different. 

https://superblog.ai/blog/using-prisma-with-nextjs-at-scale-ckpld2icd05251yo0mjow4fmi

**Comment by janpio** [comment]
@gregorcee123 @matt-development-work Are you also using `migrate dev` when you get this?
If not, please open a new issue and provide all the information you have so we can understand better. Thanks.

**Comment by simonv3** [comment]
I'm getting this with `npx prisma migrate dev`. This just started happening to me after stepping away from the project I was using Prisma for.

```
$ docker exec -it nomads-api yarn migrate
yarn run v1.22.19
$ DEBUG=* npx prisma migrate dev
  prisma:engines  binaries to download libquery-engine, migration-engine, introspection-engine, prisma-fmt +0ms
  prisma:cache-dir  The following error is being caught and just there for debugging: +0ms
  prisma:cache-dir  [Error: EACCES: permission denied, mkdir '/root/.cache/prisma/master/d6e67a83f971b175a593ccc12e15c4a757f93ffe/linux-musl'] {
  errno: -13,
  code: 'EACCES',
  syscall: 'mkdir',
  path: '/root/.cache/prisma/master/d6e67a83f971b175a593ccc12e15c4a757f93ffe/linux-musl'
} +0ms
  prisma:cache-dir  The following error is being caught and just there for debugging: +3ms
  prisma:cache-dir  [Error: EACCES: permission denied, mkdir '/root/.cache/prisma/master/d6e67a83f971b175a593ccc12e15c4a757f93ffe/linux-musl'] {
  errno: -13,
  code: 'EACCES',
  syscall: 'mkdir',
  path: '/root/.cache/prisma/master/d6e67a83f971b175a593ccc12e15c4a757f93ffe/linux-musl'
} +0ms
  prisma:cache-dir  The following error is being caught and just there for debugging: +11ms
  prisma:cache-dir  [Error: EACCES: permission denied, mkdir '/root/.cache/prisma/master/d6e67a83f971b175a593ccc12e15c4a757f93ffe/linux-musl'] {
  errno: -13,
  code: 'EACCES',
  syscall: 'mkdir',
  path: '/root/.cache/prisma/master/d6e67a83f971b175a593ccc12e15c4a757f93ffe/linux-musl'
} +0ms
  prisma:cache-dir  The following error is being caught and just there for debugging: +5ms
  prisma:cache-dir  [Error: EACCES: permission denied, mkdir '/root/.cache/prisma/master/d6e67a83f971b175a593ccc12e15c4a757f93ffe/linux-musl'] {
  errno: -13,
  code: 'EACCES',
  syscall: 'mkdir',
  path: '/root/.cache/prisma/master/d6e67a83f971b175a593ccc12e15c4a757f93ffe/linux-musl'
} +0ms
  prisma:loadEnv  project root found at /var/www/api/package.json +0ms
  prisma:tryLoadEnv  Environment variables loaded from /var/www/api/.env +0ms
  prisma:getConfig  Using getConfig Wasm +0ms
  prisma:getConfig  config data retrieved without errors in getConfig Wasm +5ms
  prisma:loadEnv  project root found at /var/www/api/package.json +13ms
  prisma:tryLoadEnv  Environment variables loaded from /var/www/api/.env +12ms
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
  prisma:getConfig  Using getConfig Wasm +4ms
  prisma:getConfig  config data retrieved without errors in getConfig Wasm +3ms
Datasource "db": PostgreSQL database "nomads", schema "public" at "pgsql:5432"

  prisma:getDMMF  Using CLI Query Engine (Node-API Library) at: /var/www/api/node_modules/@prisma/engines/libquery_engine-linux-musl.so.node +0ms
  prisma:getDMMF  Loaded Node-API Library +0ms
  prisma:getDMMF  unserialized dmmf result ready +18ms
  prisma:getDMMF  dmmf retrieved without errors in getDmmfNodeAPI +5ms
  prisma:getConfig  Using getConfig Wasm +50ms
  prisma:getConfig  config data retrieved without errors in getConfig Wasm +2ms
  prisma:getConfig  Using getConfig Wasm +1ms
  prisma:getConfig  config data retrieved without errors in getConfig Wasm +1ms
  prisma:migrateEngine:rpc  starting migration engine with binary: /var/www/api/node_modules/@prisma/engines/migration-engine-linux-musl +0ms
  prisma:migrateEngine:rpc  SENDING RPC CALL {"id":1,"jsonrpc":"2.0","method":"devDiagnostic","params":{"migrationsDirectoryPath":"/var/www/api/prisma/migrations"}} +5ms
  prisma:migrateEngine:stderr  {"timestamp":"2023-03-02T21:12:13.248862382Z","level":"INFO","fields":{"message":"Starting migration engine RPC server","git_hash":"d6e67a83f971b175a593ccc12e15c4a757f93ffe"},"target":"migration_engine"} +0ms
  prisma:migrate:dev  {
  devDiagnostic: '{\n  "action": {\n    "tag": "createMigration"\n  }\n}'
} +0ms
  prisma:migrateEngine:rpc  SENDING RPC CALL {"id":2,"jsonrpc":"2.0","method":"applyMigrations","params":{"migrationsDirectoryPath":"/var/www/api/prisma/migrations"}} +764ms
  prisma:migrateEngine:rpc  {
  jsonrpc: '2.0',
  error: {
    code: 4466,
    message: 'An error happened. Check the data field for details.',
    data: {
      is_panic: false,
      message: 'The database server at `pgsql`:`5432` was reached but timed out.\n' +
        '\n' +
        'Please try again.\n' +
        '\n' +
        'Please make sure your database server is running at `pgsql`:`5432`.\n' +
        '\n' +
        'Context: Timed out trying to acquire a postgres advisory lock (SELECT pg_advisory_lock(72707369)). Elapsed: 10000ms. See https://pris.ly/d/migrate-advisory-locking for details.\n',
      meta: [Object],
      error_code: 'P1002'
    }
  },
  id: 2
} +10s
Error: Error: P1002

The database server at `pgsql`:`5432` was reached but timed out.

Please try again.

Please make sure your database server is running at `pgsql`:`5432`.

Context: Timed out trying to acquire a postgres advisory lock (SELECT pg_advisory_lock(72707369)). Elapsed: 10000ms. See https://pris.ly/d/migrate-advisory-locking for details.


    at Object.<anonymous> (/var/www/api/node_modules/prisma/build/index.js:93743:25)
    at MigrateEngine.handleResponse (/var/www/api/node_modules/prisma/build/index.js:93596:36)
    at LineStream3.<anonymous> (/var/www/api/node_modules/prisma/build/index.js:93695:16)
    at LineStream3.emit (node:events:513:28)
    at addChunk (node:internal/streams/readable:324:12)
    at readableAddChunk (node:internal/streams/readable:297:9)
    at Readable.push (node:internal/streams/readable:234:10)
    at LineStream3._pushBuffer (/var/www/api/node_modules/prisma/build/index.js:93473:17)
    at LineStream3._transform (/var/www/api/node_modules/prisma/build/index.js:93467:8)
    at Transform._write (node:internal/streams/transform:175:8)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

I've got @prisma/client@4.11.0. This is all running inside a docker setup [which you can see here](https://github.com/funmusicplace/nomads/blob/main/docker-compose.yml).

**Comment by alexandrosk** [comment]
@simonv3 Same problem here, exactly like you described. :hmmm:

**Comment by amiralam0** [comment]
same

**Comment by frankda** [comment]
Resolved by enable [PRISMA_SCHEMA_DISABLE_ADVISORY_LOCK](https://www.prisma.io/docs/reference/api-reference/environment-variables-reference#prisma_disable_advisory_lock). Guess it is a network issue

**Comment by amirbenasr** [comment]
> Resolved by enable [PRISMA_SCHEMA_DISABLE_ADVISORY_LOCK](https://www.prisma.io/docs/reference/api-reference/environment-variables-reference#prisma_disable_advisory_lock). Guess it is a network issue

can you explain more ?

**Comment by DibPuelma** [comment]
> Resolved by enable [PRISMA_SCHEMA_DISABLE_ADVISORY_LOCK](https://www.prisma.io/docs/reference/api-reference/environment-variables-reference#prisma_disable_advisory_lock). Guess it is a network issue

This worked for me, don't know why though

**Comment by CharlesCCC** [comment]
> Resolved by enable [PRISMA_SCHEMA_DISABLE_ADVISORY_LOCK](https://www.prisma.io/docs/reference/api-reference/environment-variables-reference#prisma_disable_advisory_lock). Guess it is a network issue

Thank you , this worked. 

For local, you can add the value in your `.env` file 
`PRISMA_SCHEMA_DISABLE_ADVISORY_LOCK=true` 

For gitlab pipeline, you can update `.gitlab-ci.yml` with
```
    variables:
        PRISMA_SCHEMA_DISABLE_ADVISORY_LOCK: true   
```

**Comment by BackstromForsberg** [comment]
> For local, you can add the value in your .env file
PRISMA_SCHEMA_DISABLE_ADVISORY_LOCK=true

Confirming that this is still working in 2024. Helped unblock my dockerized local development issues, running:
`docker exec container-name npx prisma migrate dev --name latest`

The weird thing is that it ran totally fine a few times before without needing this `.env` value, but now won't work without it.

**Comment by AdiCahyaSaputra** [comment]
I think when you run the `prisma migrate ...` command and failed or hang up, then you retry to run migrate again, prisma will block it (make it timeout). You can set the `PRISMA_SCHEMA_DISABLE_ADVISORY_LOCK` to byass the lock

[updated reference for prisma ^5.3.0](https://www.prisma.io/docs/orm/prisma-migrate/workflows/development-and-production#advisory-locking)

[you can manually unlock it](https://medium.com/israeli-tech-radar/fixing-a-stuck-postgres-advisory-lock-without-restarting-your-db-02c33d0f0c08)
